frontend/src/screens/ProductScreen.js

```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { Link } from 'react-router-dom';
+import Rating from '../components/Rating';
+import data from '../data';
+
+export default function ProductScreen(props) {
+  const product = data.products.find((x) => x._id === props.match.params.id);
+  if (!product) {
+    return <div> Product Not Found</div>;
+  }
+  return (
+    <div>
+      <Link to="/">Back to result</Link>
+      <div className="row top">
+        <div className="col-2">
+          <img className="large" src={product.image} alt={product.name}></img>
+        </div>
+        <div className="col-1">
+          <ul>
+            <li>
+              <h1>{product.name}</h1>
+            </li>
+            <li>
+              <Rating
+                rating={product.rating}
+                numReviews={product.numReviews}
+              ></Rating>
+            </li>
+            <li>Pirce : ${product.price}</li>
```

Comment:
"Pirce" should be "Price"

Reply:
![Capture](https://user-images.githubusercontent.com/46791411/142716544-7b65b9df-17fb-4497-a852-e493a8fcefc9.PNG)

what do I need to change for this to work

Reply:
Hi,
it seems you just need to remove "this" so the line 7 will be:
 const product = data.products.find( (x) => x._id === props.match.params.id)
 
 We are using this keyword only with class components. 
 Hope it helps

Reply:
Since react-router v5.1 with hooks:

import { useParams } from 'react-router';

And instead of 
this.props.match.params.id
you should just use this hook like below:

  const { id } = useParams();
  const product = data.products.find((x) => x._id === id);

And as well in App.js you should wrap Route with Routes like this:
 <Routes>
            <Route path="/" element={<HomeScreen />} exact />
            <Route path="/product/:id" element={<ProductScreen />} />
 </Routes>

of course you need to import extra Routes from "react-router-dom"
He is using useParams later in a course anyway.

Hope this helps

Reply:
I try your solution and it shows me this error.  It looks like component didn't connect with data.js

TypeError: Cannot read properties of undefined (reading 'image')
 15 |   <div className="col-2">
> 16 |     <img className="large" src={product.image} alt={product.name}></img>
     | ^  17 |   </div>

Reply:
15 |   <div className="col-2">
> 16 |     <img className="large" src={product.image} alt={product.name}></img>
     | ^  17 |   </div>

Reply:
` 15 |   <div className="col-2">
> 16 |     <img className="large" src={product.image} alt={product.name}></img>
     | ^  17 |   </div>`

Reply:
![Screenshot 2021-11-28 at 15 14 42](https://user-images.githubusercontent.com/44552832/143773987-7163dc11-fe7c-4271-b52c-f621184cbc1c.png)

Reply:
Hi,
This looks like your product is undefined.
You may check if is properly selected from data.js using find() and I assume you probably did.
Try to console.log(product) and play it out till you get demanded response to say more I'll need to see whole ProductScreen.js and App.js. You may update and push your repo which let me see your code.

Let me know if you either you uploading or have found solution yourself as this might help others while version of Router has changed and it does little confusion

Reply:
All the other files are according to the video lectures.
I declare routes like this then its works, 
`    <Routes>
      <Route path="/" element={<HomeScreen /> } > </Route>
      </Routes>`
product is undefined, if I declare any number to product const, it shows me that number, I think there is some issue with the UseParams  
This is my code: 

`import React from 'react';
import { Link } from 'react-router-dom';
import Rating from '../Rating'
import data from '../data';
import { useParams } from 'react-router';
export default function ProductScreen() {
const { id } = useParams();
const product = data.products.find((x) => x._id === id);
console.log(product); // undefined
  return (
    <div>
      <Link to="/">Back to result</Link>
      <div className="row top">
        <div className="col-2">
          <img className="large" src={product.image} alt={product.name}></img>
        </div>`

Reply:
Hi,
Maybe your id is undefined and that's what makes const product undefined as well? 
log id and see if contains any value.

useParams does take id from query string so it must be passed in address bar like this:
localhost:3000/api/product/3 where number 3 is and product id and it exists in data.js.

File ProductScreen should be added into Routes in App.js as well.

You should try and follow rest of tutorial as your code looks incomplete but I can not see it all so it is hard to say for sure.
Also looks like Route for ProductScreen is missing in App.js as well.

Hope it somehow helps

Reply:
@iamjacek as you said I continued the course, I am now at the productScreen page using redux at /pull/13/files. 
code is same,  I get this error "
TypeError: Cannot read properties of undefined (reading 'params')" in productScreen file.
` 
>const dispatch = useDispatch();
> 24 | const productId = props.match.params.id;`

I also followed the steps you mention above, about this same error

Reply:
![error](https://user-images.githubusercontent.com/80596000/145431781-1ed692b1-bc0d-46cb-8b16-538b1000ef67.png)

Reply:
![error2](https://user-images.githubusercontent.com/80596000/145434250-4f3e3dd6-c5ea-4193-9e53-e0cf69f44e2f.png)

Reply:
> 

This really helped .. thanks a ton.